torch/utils/data/datapipes/iter/combining.py

```diff
@@ -196,25 +193,42 @@ class _ChildDataPipe(IterDataPipe):
     Iterable Datapipe that is a child of a main DataPipe. The instance of this class
     will pass its instance_id to get the next value from its main DataPipe.
 
+    Note:
+        ChildDataPipe, like all other IterDataPipe, follows the singler iterator per IterDataPipe constraint.
+        Since ChildDataPipes share a common buffer, when an iterator is created for one of the ChildDataPipes,
+        the previous iterators  for all ChildDataPipes must be invalidated, with the exception when a ChildDataPipe
+        hasn't had an iterator created from it since the last invalidation. See the example below.
+
+    Singler Iterator per IteraDataPipe Invalidation Example:
+        >>> source_dp = IterableWrapper(range(10))
+        >>> cdp1, cdp2 = source_dp.fork(num_instances=2)
+        >>> it1, it2 = iter(cdp1), iter(cdp2)
+        >>> it3 = iter(cdp1)
+        The line above invalidates `it1` and `it2`, and resets `ForkerIterDataPipe`.
+        >>> it4 = iter(cdp2)
+        The line above doesn't invalidate `it3`, because an iterator for `cdp2` hasn't been created since
+        the last invalidation.
+
     Args:
         main_datapipe: Main DataPipe with a method 'get_next_element_by_instance(instance_id)'
         instance_id: integer identifier of this instance
     """
-    def __init__(self, main_datapipe, instance_id: int):
-        required_attrs = ["get_next_element_by_instance", "is_instance_started", "is_every_instance_exhausted", "reset"]
+    _is_child_datapipe: bool = True
+
+    def __init__(self, main_datapipe: IterDataPipe, instance_id: int):
+        required_attrs = ["get_next_element_by_instance", "is_every_instance_exhausted", "reset"]
         required_ops = [getattr(main_datapipe, attr) for attr in required_attrs]
         if any(not callable(op) for op in required_ops):
             raise NotImplementedError(f"Main Datapipe must have methods {required_attrs} implemented.")
-        self.main_datapipe = main_datapipe
+        self.main_datapipe: IterDataPipe = main_datapipe
         self.instance_id = instance_id
 
     def __iter__(self):
-        if self.main_datapipe.is_instance_started(self.instance_id):  # Only reset if the DataPipe started to read
-            if not self.main_datapipe.is_every_instance_exhausted():
-                warnings.warn("Some child DataPipes are not exhausted when __iter__ is called. We are resetting "
-                              "the buffer and each child DataPipe will read from the start again.", UserWarning)
-            self.main_datapipe.reset()
-        # We want to separate the code for reset and yield, so that 'reset' exeutes before __next__ is called
+        if not self.main_datapipe.is_every_instance_exhausted():
+            warnings.warn("Some child DataPipes are not exhausted when __iter__ is called. We are resetting "
+                          "the buffer and each child DataPipe will read from the start again.", UserWarning)
```

Comment:
This is currently causing a lot of warnings in our tests. It may causes too many warnings for users too if they happen to have the need to reset before reading everything. I think we have the following options:
1. Remove the warning entirely
2. Raises once, then not again (have a global variable that switches)
3. Option to silence this through a global API?
4. Keep this as it is

Reply:
I can confirm through the tests within `torchvision` that it raises a lot of warning, so we definitely should not keep it as it is.

Reply:
I support to silence all warnings within TorchData via a global API. For users, this warning is good to indicate their wrong implementation or some potential bugs because this won't happen a lot. But, for `torchvision` tests, it's reasonable to disable it as the test was introduced intentionally.

Reply:
IIUC, this warning should only pop up if we actually are doing something dubious in the first place, right? @NivekT what datasets are giving you this warning? Should we simply fix our datasets instead of fixing the warning behavior?

Reply:
> Raises once, then not again (have a global variable that switches)

This is the default Python behavior: 

```py
import warnings

def warn():
    warnings.warn("You should only see this once!")

for _ in range(5):
    warn()
```

The reason you are seeing this warning over and over in our test suite is that `pytest` does something similar to `warnings.simplefilter("always")`.

Reply:
I re-test `torchvision` with my latest changes (after I moved the warning to the correct location). I'm no longer seeing any warnings.